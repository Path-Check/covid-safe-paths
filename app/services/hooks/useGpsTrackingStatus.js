import BackgroundGeolocation from '@mauron85/react-native-background-geolocation';
import { useEffect, useState } from 'react';

import LocationService from '../LocationService';

/**
 * Hook used to check the location tracking status of a user
 * @param {{canTrack: boolean, reason: Reason, isRunning: boolean}} status
 */
export const useGpsTrackingStatus = () => {
  const initialState = {
    canTrack: undefined,
    reason: undefined,
    isRunning: undefined,
  };

  const [userOptIn, setStatus] = useState(initialState);

  /**
   * Set the location tracking status, and start/stop BackgroundGeolocation accordingly
   * @param {boolean} userOptIn
   * @returns {{canTrack: boolean, reason: Reason, isRunning: boolean}}
   */
  const setUserOptIn = async userOptIn => {
    LocationService.setUserOptIn(userOptIn);

    const { isRunning } = await LocationService.checkStatus();
    const shouldStartTracking = userOptIn && !isRunning;

    shouldStartTracking
      ? await LocationService.start()
      : await LocationService.stop();

    const newStatus = await LocationService.checkStatus();

    setStatus(newStatus);
  };

  useEffect(() => {
    BackgroundGeolocation.on('start', async () => {
      await getAndSetStatus();
    });

    const getAndSetStatus = async () => {
      const status = await LocationService.checkStatus();
      setStatus(status);
    };

    getAndSetStatus();
  }, []);

  return [userOptIn, setUserOptIn];
};
