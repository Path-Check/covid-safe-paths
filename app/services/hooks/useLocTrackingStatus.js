import BackgroundGeolocation from '@mauron85/react-native-background-geolocation';
import { useEffect, useState } from 'react';

import LocationService from '../LocationService';

/**
 * Hook used to check the location tracking status of a user
 * @param {{canTrack: boolean, reason: Reason, isRunning: boolean, hasPotentialExposure: boolean}} status
 */
export const useLocTrackingStatus = () => {
  const initialState = {
    canTrack: undefined,
    reason: undefined,
    isRunning: undefined,
    hasPotentialExposure: undefined,
  };

  const [locTrackingStatus, setStatus] = useState(initialState);

  /**
   * Set the location tracking status, and start/stop BackgroundGeolocation accordingly
   * @param {boolean} newIsTracking
   * @returns {{canTrack: boolean, reason: Reason, isRunning: boolean, hasPotentialExposure: boolean}}
   */
  const setLocTrackingStatus = async newIsTracking => {
    LocationService.setStoreLocationTracking(newIsTracking);

    const { isRunning } = await LocationService.checkStatus();
    const shouldStartTracking = newIsTracking && !isRunning;

    shouldStartTracking
      ? await LocationService.start()
      : await LocationService.stop();

    const newStatus = await LocationService.checkStatus();

    setStatus(newStatus);
  };

  useEffect(() => {
    BackgroundGeolocation.on('start', async () => {
      await getAndSetStatus();
    });

    const getAndSetStatus = async () => {
      const status = await LocationService.checkStatus();
      setStatus(status);
    };

    getAndSetStatus();
  }, []);

  return [locTrackingStatus, setLocTrackingStatus];
};
