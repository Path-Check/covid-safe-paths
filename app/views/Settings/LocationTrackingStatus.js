import styled, { css } from '@emotion/native';
import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Linking } from 'react-native';
import DeviceSettings from 'react-native-device-settings';
import { openSettings } from 'react-native-permissions';

import { Button, Divider, Switch, Typography } from '../../components';
import Colors from '../../constants/colors';
import { config } from '../../COVIDSafePathsConfig';
import { useLocTrackingStatus } from '../../services/hooks/useLocTrackingStatus';
import { Reason } from '../../services/LocationService';
import { isPlatformiOS } from '../../Util';

export const LocationTrackingStatus = () => {
  const { i18n } = useTranslation();
  const [locTrackingStatus, setLocTrackingStatus] = useLocTrackingStatus();

  const { reason, canTrack } = locTrackingStatus;

  const isGPS = config.tracingStrategy === 'gps';

  console.log(reason);

  const hasLocTrackingPermissions =
    reason === Reason.USER_ENABLED || reason === Reason.USER_DISABLED;

  const isMissingLocPermissions =
    reason === Reason.DEVICE_LOCATION_OFF ||
    reason === Reason.APP_NOT_AUTHORIZED;

  const getCurrentRowDirection = () =>
    i18n.dir() === 'rtl' ? 'row-reverse' : 'row';

  const getLocStatusElement = () => {
    if (hasLocTrackingPermissions) {
      return (
        <LocationStatusSwitch
          canTrack={canTrack}
          setLocTrackingStatus={setLocTrackingStatus}
        />
      );
    }

    if (isMissingLocPermissions) {
      return <EnableLocationButton reason={reason} />;
    }
  };

  if (isGPS) {
    return (
      <>
        <Container
          style={css`
            flex-direction: ${getCurrentRowDirection()};
          `}>
          {getLocStatusElement()}
        </Container>
        <Divider />
      </>
    );
  }
};

const EnableLocationButton = ({ reason }) => {
  const { t } = useTranslation();

  /**
   * Conditional toggling based on the location permission actions
   * that are enabled per platform.
   */
  const requestLocationPermission = async () => {
    if (reason === Reason.APP_NOT_AUTHORIZED) {
      /**
       * Open the settings for the app
       */
      openSettings();
    } else if (reason === Reason.DEVICE_LOCATION_OFF) {
      /**
       * Open Device Settings.
       * If iOS, open Settings home screen
       * If Android, we are able to open the device location permissions
       */
      if (isPlatformiOS()) {
        const deviceSettingsUrl = 'App-prefs:';
        Linking.openURL(deviceSettingsUrl);
      } else {
        DeviceSettings.location();
      }
    }
  };

  return (
    <>
      <Button
        onPress={requestLocationPermission}
        label={t('label.logging_request_location_permission')}
        style={{ width: '100%' }}
        secondary
        small
      />
    </>
  );
};

const LocationStatusSwitch = ({ canTrack, setLocTrackingStatus }) => {
  const { t, i18n } = useTranslation();
  const [label, setLabel] = useState('');

  useEffect(() => {
    let labelInterval;

    /**
     * Provide a basic animation to show that location logging
     * is active. If the label reaches three periods, reset
     * to the original label
     */
    const createLabelInterval = () => {
      return setInterval(() => {
        setLabel(label => {
          if (label.slice(-3) === '...') {
            setLabel(t('label.logging_active_location'));
          } else {
            setLabel(label + '.');
          }
        });
      }, 1000);
    };

    const initLabel = () => {
      if (canTrack) {
        setLabel(t('label.logging_active_location'));
        labelInterval = createLabelInterval();
      } else {
        setLabel(t('label.logging_location_paused'));
        labelInterval = clearInterval(labelInterval);
      }
    };

    initLabel();

    return () => clearInterval(labelInterval);
  }, [canTrack, t]);

  const getCurrentMarginDirection = () =>
    i18n.dir() === 'rtl' ? 'margin-left: 12px;' : 'margin-right: 12px;';

  const onValueChange = async val => await setLocTrackingStatus(val);

  return (
    <>
      <Label>
        <Typography use='body1' style={!canTrack && { color: Colors.RED_TEXT }}>
          {label}
        </Typography>
      </Label>
      <Switch
        onValueChange={onValueChange}
        value={canTrack}
        style={css`
          ${getCurrentMarginDirection()};
        `}
      />
    </>
  );
};

const Container = styled.View`
  padding: 18px 0;
  align-items: center;
`;

const Label = styled.View`
  flex: 1;
  justify-content: center;
`;
