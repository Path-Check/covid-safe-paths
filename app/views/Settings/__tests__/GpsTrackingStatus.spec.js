import { act, render } from '@testing-library/react-native';
import React from 'react';

// import { config } from '../../../COVIDSafePathsConfig';
import * as useGpsTrackingStatus from '../../../services/hooks/useGpsTrackingStatus';
import { LocationTrackingStatus } from '../LocationTrackingStatus';

jest.useFakeTimers();

describe('<LocationTrackingStatus />', () => {
  describe('when contact tracing strategy is not GPS', () => {
    jest.mock('../../../COVIDSafePathsConfig', () => ({
      config: {
        tracingStrategy: '',
      },
    }));

    it('does not render', () => {
      const { asJSON } = render(<LocationTrackingStatus />);
      expect(asJSON()).toMatchSnapshot();
    });
  });

  describe('when contact tracing strategy is GPS', () => {
    jest.mock('../../../COVIDSafePathsConfig', () => ({
      config: {
        tracingStrategy: 'gps',
      },
    }));

    it('renders a <LocationStatusSwitch /> if hasPermission', async () => {
      jest
        .spyOn(useGpsTrackingStatus, 'useGpsTrackingStatus')
        .mockReturnValueOnce([
          {
            reason: 'ALL_CONDITIONS_MET',
            canTrack: true,
          },
          jest.fn(),
        ]);

      const { asJSON } = render(<LocationTrackingStatus />);

      await act(async () => {
        jest.runAllTimers();
      });

      await expect(asJSON()).toMatchSnapshot();
    });
    it('renders a <EnableLocationButton /> if isMissingPermission', () => {});
  });
});
