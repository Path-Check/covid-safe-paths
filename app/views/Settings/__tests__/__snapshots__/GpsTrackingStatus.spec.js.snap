// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationTrackingStatus /> when contact tracing strategy is GPS renders a <LocationStatusSwitch /> if hasPermission 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`<LocationTrackingStatus /> when contact tracing strategy is not GPS does not render 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
